File: py/pe/pe1.py
#!/usr/bin/env python
"""
Solve project euler 1

Add all the natural numbers below one thousand
that are multiples of 3 or 5.
"""
from __future__ import division, print_function
#unicode_literals, absolute_import

def pe1a(limit=1000):
    """
    iter

    >>> pe1a()
    234168
    """
    return sum([x for x in range(1, limit + 1) if not x % 3 or not x % 5])

def pe1(n=1000):
    """
    sum(multiples of 3) + sum(multiples of 5) - sum(multiples of 15)

    >>> pe1()
    234168
    """
    s = 0

    if n > 2:
        n3 = n // 3
        n5 = n // 5
        n15 = n // 15
        s = (3 * n3 * (n3 + 1) + 5 * n5 * (n5 + 1) - 15 * n15 * (n15 + 1)) >> 1

    return s

if __name__ == "__main__":
    import doctest
    doctest.testmod()

    try:
        while True:
            s = input('> ')
            n = int(s)
            print(pe1(n))
    except (SyntaxError, EOFError, KeyboardInterrupt, NameError):
        pass

File: rs/src/pe/pe1.rs
/// # Solve project euler 1
///
/// Add all the natural numbers below one thousand
/// that are multiples of 3 or 5.

fn pe1a(n: u32) -> u32 {
    let mut s: u32 = 0;
    for i in 1..n + 1 {
        if i % 3 == 0 || i % 5 == 0 {
            s += i;
        }
    }
    return s;
}

fn pe1b(n: u32) -> u32 {
    let mut s: u32 = 0;
    if n > 2 {
        let n3: u32 = n / 3;
        let n5: u32 = n / 5;
        let n15: u32 = n / 15;
        s = (3 * n3 * (n3 + 1) + 5 * n5 * (n5 + 1) - 15 * n15 * (n15 + 1)) >> 1;
    }
    return s;
}

pub fn pe1() {
    let s: u32 = pe1a(1000);
    println!("{}", s);
}
